### PR TITLE
Stop masking errors in channel lookups

### PR DESCRIPTION
**Pull request by jerith**
Currently, any API call that requires a channel (or a group or a user) starts by making a lookup call. In this call, any failure response is silently replaced with `channel_not_found` (or `user_not_found`) which leads to much confusion when trying to figure out why something isn't working.

For `lookupk` to pass through all the errors it gets without running into type trouble, all the list functions it's passed must have the same set of errors in their return types. I checked slack's API documentation to see if `groups_list` actual needs the `bot_error` in its signature. Both https://api.slack.com/methods/groups.list and https://api.slack.com/bot-users indicate that it doesn't, so I removed it. (I did the same for `im_list` even it isn't used with `lookupk`, because it probably should be at some point.)

There's probably some more cleanup that could be done here, but I didn't want to change too much before getting a review. :-)

### CODE REVIEW

File: src/slacko.ml [hunk 1]
```diff
@@ -765,33 +765,22 @@ let groups_list ?exclude_archived session =
...
 (* look up the id of query from results provided by the listfn *)
 let lookupk session listfn filterfn k =
   match%lwt listfn session with
-  | `Success channels -> (match List.filter filterfn channels with
-    | [] -> Lwt.fail No_matches
-    | [x] -> Lwt.return @@ k x
-    | _ -> Lwt.fail No_unique_matches)
-  | _ -> Lwt.fail Lookup_failed
+  | #parsed_auth_error as e -> Lwt.return e
+  | `Success items -> Lwt.return @@ k @@ List.filter filterfn items
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: What if the lookup returns neither an `auth_error` nor a `Success`? It should probably have an `| _ -> ???` case with some defined behaviour.

I am not sure whether taking the "ensure there is one channel" logic out from here and the duplicaing it in every caller is a good idea. They all need to handle the case in probably the same way, so it might as well be done in one, common location.

File: src/slacko.ml [hunk 2]
```diff
@@ -765,33 +765,22 @@ let groups_list ?exclude_archived session =
...
   | ChannelName name ->
     let base = String.sub name 1 @@ String.length name - 1 in
-    match%lwt (lookupk session channels_list (fun (x:channel_obj) -> x.name = base) @@ function
-    | {id = ChannelId s; _} -> Some s
-    | {id = ChannelName _; _} -> None) with
-    | exception _ -> Lwt.return `Channel_not_found
-    | Some v -> Lwt.return @@ `Found v
-    | None -> Lwt.return `Channel_not_found
+    lookupk session channels_list (fun (x:channel_obj) -> x.name = base) @@ function
+    | [{id = ChannelId s; _}] -> `Found s
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: What if we get less or more than exactly one hit? That should not happen, but I wouldn't want the pattern match to blow up.

File: src/slacko.ml [hunk 3]
```diff
@@ -801,24 +790,18 @@ let string_of_channel = function
...
   | UserName name ->
-    match%lwt (lookupk session users_list (fun (x:user_obj) -> x.name = name) @@ function
-    | {id = UserId s; _} -> Some s
-    | {id = UserName _; _} -> None) with
-    | exception _ -> Lwt.return `User_not_found
-    | Some v -> Lwt.return @@ `Found v
-    | None -> Lwt.return `User_not_found
+    lookupk session users_list (fun (x:user_obj) -> x.name = name) @@ function
+    | [{id = UserId s; _}] -> `Found s
+    | _ -> `User_not_found
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: Same as above.

File: src/slacko.ml [hunk 4]
```diff
@@ -801,24 +790,18 @@ let string_of_channel = function
...
-    | {id = GroupId s; _} -> Some s
-    | {id = GroupName _; _} -> None) with
-    | exception _ -> Lwt.return `Channel_not_found
-    | Some v -> Lwt.return @@ `Found v
-    | None -> Lwt.return `Channel_not_found
-
-let id_of_chat session : _ -> [ `Found of string | `Channel_not_found | `User_not_found ] Lwt.t = function
+    lookupk session groups_list (fun (x:group_obj) -> x.name = name) @@ function
+    | [{id = GroupId s; _}] -> `Found s
+    | _ -> `Channel_not_found
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: Same as above.

File: src/slacko.ml (hunk 1 above)
Comment (jerith) [inline src/slacko.ml]: The first thing is all about types, and this is one of those cases where it's probably better to make them explicit instead of inferring them. I'll do that, but first I'll explain what's going on.

TL;DR: The pattern match must be exhaustive or the compile fails with a type error.

Here's the inferred type of `lookupk` (according to merlin):
```
'a ->
('a ->
 [< `Account_inactive
  | `Invalid_auth
  | `Not_authed
  | `ParseFailure of string
  | `Success of 'b list
  | `Unhandled_error of string
  | `Unknown_error ]
 Lwt.t) ->
('b -> bool) -> ('b list -> ([> parsed_auth_error ] as 'c)) -> 'c Lwt.t
```
The first arg is a `session`, but we don't care as long as we can pass it to `listfn`. The second arg is `listfn`. We give it the session (`'a`) and we `match%lwt` its return value, so its return value is inferred to be (a subset of) the variants we know how to handle (wrapped in `Lwt.t`). That gives us the set of allowed errors (from the `| #parsed_auth_error as e ->` match) and `` `Success of 'b list ``. We don't care what `'b` is, as long as we can pass it to `filterfn` (the third arg with type `'b -> bool`) and `k` (the fourth arg).

`k` is inferred to have the type `'b list -> ([> parsed_auth_error ] as 'c)`, where `'c` is also `lookupk`'s return type` because we return whatever value it gives us. Because we can also return a `parsed_auth_error` from the other match, `k`'s return type must be compatible with that. However, `'c` includes the "not found" error value (either `` `Channel_not_found `` or `` `User_not_found ``) and that must remain opaque to `lookupk` so we can use it for both channels and users without adding superfluous errors to their return types.

I can (and plan to) define `type 'a listfn = session -> [`Success of 'a list | parsed_auth_error] Lwt.t` and annotate `lookupk`'s `listfn` arg to make it explicit (and make the inferred types more readable), but that's essentially what's being inferred already anyway.

Comment (jerith) [inline src/slacko.ml]: I bounced back and forth a few times on the "exactly one result" thing. While it belongs in `lookupk` conceptually, putting it there just adds a pile of extra values to pass around. If we handle "too many" and "too few" cases separately, we get this:
```ocaml
let lookupk session listfn filterfn k =
  match%lwt listfn session with
  | #parsed_auth_error as e -> Lwt.return e
  | `Success items -> match List.filter filterfn items with
    | [item] -> Lwt.return @@ k (`Found item)
    | [] -> Lwt.return @@ k (`Too_few)
    | _::_::_ -> Lwt.return @@ k (`Too_many)

let id_of_channel session = function
  | ChannelId id -> Lwt.return @@ `Found id
  | ChannelName name ->
    let base = String.sub name 1 @@ String.length name - 1 in
    lookupk session channels_list (fun (x:channel_obj) -> x.name = base) @@ function
    | `Found {id = ChannelId s; _} -> `Found s
    | `Found {id = ChannelName _; _} -> `Channel_not_found
    | `Too_few -> `Channel_not_found
    | `Too_many -> `Channel_not_found
```

We could reduce this to a pair of "found"/"not found" cases and pass a `'b option` to `k`, but we'd still be splitting the result into multiple parts and then recombining them in `k`. We could also return a "not found" error directly from `lookupk`, but then we'd need to pass that in (because it's different for different callers) and we'd still need to handle the `<thing>Id` vs `<thing>Name` split in the caller anyway.

I suppose we could merge `filterfn` and `k` into a single function that returns `Some s` for a matching object with an appropriate id and `None` for a mismatch or no id, but we'd still have to pass in the "not found value. I'll try that and see if it ends up being cleaner.

File: src/slacko.ml (hunk 2 above)
Comment (jerith) [inline src/slacko.ml]: The match won't blow up, because the `` | _ -> `Channel_not_found `` case will catch any list that doesn't contain exactly one item.

File: src/slacko.ml (hunk 1 above)
Comment (jerith) [inline src/slacko.ml]: That ends up looking like this:
```ocaml
let filter_map f l =
  List.fold_left
    (fun xs x -> match f x with None -> xs | Some v -> v :: xs) [] l

(* look up the id of query from results provided by the listfn *)
let lookupk session (listfn : 'a listfn) notfound filterfn =
  match%lwt listfn session with
  | #parsed_auth_error as e -> Lwt.return e
  | `Success items -> match filter_map filterfn items with
    | [id] -> Lwt.return @@ `Found id
    | _ -> Lwt.return notfound

let id_of_channel session = function
  | ChannelId id -> Lwt.return @@ `Found id
  | ChannelName name ->
    let base = String.sub name 1 @@ String.length name - 1 in
    lookupk session channels_list `Channel_not_found @@ function
    | {name = n; id = ChannelId s; _} when n = base -> Some s
    | _ -> None
```

I don't think it's any better than any of the other options, but I'm happy to go with whichever you'd prefer.

File: src/slacko.ml (hunk 2 above)
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: Though I wouldn't say that this is the right error in this case, it rather masks the actual issue (which is most likely a problem or change with the Slack API and requires adjustments in Slacko).

File: src/slacko.ml (hunk 1 above)
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: I agree that we shouldn't add more cases like `` `Too_many`` and `` `Too_few``, since most of the time I tried for the polymorphic variants to map to errors that the Slack API actually returns, not to Slacko-internal judgements upon the return values. At least where possible.

I actually quite like the approach of merged `k` and `filterfn`, because that pulls out the match on the 1 element list into a function which is meant to know what to do with it. The only complaint I see here is that returning `notfound` whenever *anything* is wrong with the function might be masking something seriously wrong with the library, the Slack API or both. I would rather do `| [] -> Lwt.return notfound | _ -> (* throw some exception here *)` since this is a case where the user probably cannot continue in a reasonable way. I am proposing an exception here, since this is a failure case that doesn't need to be represented in the signature and the user should most likely not try to catch the exception, since it basically only says "there is a bug in Slack or Slacko".
